test/truncate.c

```diff
@@ -19,6 +19,45 @@
 #define ONE_GIG_SIZE ((loff_t)1024 * 1024 * 1024)
 #define HALF_GIG_SIZE ((loff_t)512 * 1024 * 1024)
 
+static int test_truncate(struct io_uring *ring, int fd)
+{
+	struct io_uring_cqe *cqe;
+	struct io_uring_sqe *sqe;
+	int ret;
+
+	sqe = io_uring_get_sqe(ring);
+	if (!sqe) {
+		fprintf(stderr, "get sqe failed\n");
+		goto err;
+	}
+
+	memset(sqe, 0, sizeof(*sqe));
+
+	io_uring_prep_rw(IORING_OP_FTRUNCATE, sqe, fd, "fail", 0, 4);
+
+	ret = io_uring_submit(ring);
+	if (ret <= 0) {
+		fprintf(stderr, "sqe submit failed: %d\n", ret);
+		goto err;
+	}
+
+	ret = io_uring_wait_cqe(ring, &cqe);
+	if (ret < 0) {
+		fprintf(stderr, "wait completion %d\n", ret);
+		goto err;
+	}
+	ret = cqe->res;
+	io_uring_cqe_seen(ring, cqe);
+	if (cqe->res != -EINVAL) {
+		fprintf(stderr, "unexpected truncate res %d\n", cqe->res);
+		goto err;
+	}
+
+  return 0;
```

Comment:
After this assignment, `ret = cqe->res,` the `ret` value is not used. Shouldn't it be returned?

Also, `goto err` is `return 1` while your caller checking is:
```
	ret = test_truncate(&ring, fd);
	if (ret < 0)
		goto err;
```
So `test_truncate()` will always be assumed as a success because it never returns a negative value.

Reply:
Fixed